src/lints/clone_on_copy.rs

```diff
@@ -65,3 +81,96 @@ fn check_clone_usage(
+        .expect("Failed to find semantic expression.");
+
+    // Extract the number of `@` snapshots from the type.
+    // Each `@` will later be represented as a `*` prefix in the output.
+    let (mut snapshot_count, _) = peel_snapshots(db, expr_semantic.ty());
+
+    // `clone(self: @T)` expects an `@`, so the compiler will automatically insert
+    // an `@` into the type if it was not explicitly provided by the user.
+    // In such cases, the expression will be of type `Expr::Snapshot`,
+    // meaning that `peel_snapshots` would count one coercion too many.
+    //
+    // However, if the `@` was explicitly written by the user,
+    // the expression will be of another type, such as `Expr::Var`,
+    // and `peel_snapshots` will have already counted the correct number of `@`.
+    //
+    // Therefore, we need to manually subtract one from the snapshot count
+    // when the expression is a `Expr::Snapshot` to correct this.
+    if let Expr::Snapshot(_) = expr_semantic {
+        snapshot_count -= 1;
+    };
+
+    let fixed_expr = format!(
+        "{}{}",
+        "*".repeat(snapshot_count),
+        ast_expr.as_syntax_node().get_text(db.upcast())
+    );
+
+    Some((node, fixed_expr))
+}
+
+fn get_expr_semantic(
+    db: &dyn SemanticGroup,
+    module_file_id: ModuleFileId,
+    ast_expr_binary: &ast::ExprBinary,
+) -> Option<Expr> {
+    let ast_expr = ast_expr_binary.lhs(db.upcast());
+
+    let expr_ptr = ast_expr.stable_ptr(db.upcast());
+
+    // Traverses up the syntax tree to find the nearest enclosing function (trait, impl, or free) that owns the expression.
+    // If found, retrieves the corresponding semantic expression.
+    ast_expr
+        .as_syntax_node()
+        .ancestors_with_self(db.upcast())
+        .find_map(|ancestor| {
+            let function_id = if let Some(trait_func) =
+                ast::TraitItemFunction::cast(db.upcast(), ancestor)
+            {
+                FunctionWithBodyId::Trait(
+                    TraitFunctionLongId(module_file_id, trait_func.stable_ptr(db.upcast()))
+                        .intern(db),
+                )
+            } else if let Some(func_with_body) = ast::FunctionWithBody::cast(db.upcast(), ancestor)
+            {
+                if ancestor
+                    .ancestor_of_kind(db.upcast(), SyntaxKind::ItemImpl)
+                    .is_some()
+                {
+                    FunctionWithBodyId::Impl(
+                        ImplFunctionLongId(module_file_id, func_with_body.stable_ptr(db.upcast()))
+                            .intern(db),
+                    )
+                } else {
+                    FunctionWithBodyId::Free(
+                        FreeFunctionLongId(module_file_id, func_with_body.stable_ptr(db.upcast()))
+                            .intern(db),
+                    )
+                }
+            } else {
+                return None;
+            };
```

Comment:
it's a lot of logic here, maybe we should extract it to some helper function, even in this very file?